main/EDInitMod.F90

```diff
@@ -839,151 +848,168 @@ subroutine init_cohorts( site_in, patch_in, bc_in)
 
              else
                 temp_cohort%hite        = EDPftvarcon_inst%hgt_min(pft)
-                
                 ! Calculate the plant diameter from height
                 call h2d_allom(temp_cohort%hite,pft,temp_cohort%dbh)
 
-                ! Calculate the leaf biomass from allometry
-                ! (calculates a maximum first, then applies canopy trim)
-                call bleaf(temp_cohort%dbh,pft,temp_cohort%crowndamage, &
-                     temp_cohort%canopy_trim,c_leaf)
-             end if  ! sp mode
+             endif  ! sp mode
 
-             ! Calculate total above-ground biomass from allometry
-             call bagw_allom(temp_cohort%dbh,pft,temp_cohort%crowndamage,c_agw)
+          else ! interpret as initial diameter and calculate density 
+             if(hlm_use_nocomp .eq. itrue)then
+                temp_cohort%dbh = abs(EDPftvarcon_inst%initd(pft))
 
-             ! Calculate coarse root biomass from allometry
-             call bbgw_allom(temp_cohort%dbh,pft,c_bgw)
+                ! calculate crown area of a single plant
+                dummy_n = 1.0_r8 ! make n=1 to get area of one tree
+               
+                call carea_allom(temp_cohort%dbh, dummy_n, init_spread_inventory, temp_cohort%pft, &
+                     temp_cohort%crowndamage, temp_cohort%c_area)
 
-             ! Calculate fine root biomass from allometry
-             ! (calculates a maximum and then trimming value)
-             call bfineroot(temp_cohort%dbh,pft,temp_cohort%canopy_trim,temp_cohort%l2fr,c_fnrt)
+                ! calculate initial density required to close canopy 
+                temp_cohort%n  = patch_in%area / temp_cohort%c_area
+             else
+                write(fates_log(),*) 'Negative fates_recruit_init_density can only be used in no comp mode'
+                call endrun(msg=errMsg(sourcefile, __LINE__))
+             endif
+          endif
 
-             ! Calculate sapwood biomass
-             call bsap_allom(temp_cohort%dbh,pft,temp_cohort%crowndamage, &
-                  temp_cohort%canopy_trim,a_sapw,c_sapw)
+          ! Calculate the leaf biomass from allometry
+          ! (calculates a maximum first, then applies canopy trim)
+          call bleaf(temp_cohort%dbh,pft,temp_cohort%crowndamage, &
+               temp_cohort%canopy_trim,c_leaf)
 
```

Comment:
I'm seeing non-b4b results for sp mode only regression tests for the default parameter file (i.e. with `initd = 0.2`).  I think it might be due to this change here.  The call to `bleaf` was originally being avoided for sp mode runs as part of the check starting at line 842.  Should it be moved back inside?

Reply:
Oh good catch. It does also need to be called in the no comp dbh initialisation mode though so maybe we need two calls - in both dbh and density part of the if statement starting 829?